### PR TITLE
Docs: improve contributor guide to explain how to work with tickets

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?

Part of #7013 

## Rationale for this change

Recently the question of who is / could be working on what ticket came up in discussions. For example"
* https://github.com/apache/arrow-datafusion/issues/8969#issuecomment-1910063746 from @Lordworms  about what tickets were available
* Where @Tangruilin and @simicd  ended up working on the same ticket: https://github.com/apache/arrow-datafusion/pull/8967

I think it would be good to clarify the contributing guide in this case

## What changes are included in this PR?
1. Update the contributing guide to explain the existing decorum: work on what you want, but it is polite to announce on tickets and file new tickets for new behavior

## Are these changes tested?
Docs only

### CODE REVIEW

File: docs/source/contributor-guide/index.md [hunk 1]
```diff
@@ -19,17 +19,40 @@
...
+You can find a curated [good-first-issue] list to help you get started.
+
+DataFusion is an open contribution project, and thus there is no particular
+project imposed deadline for completing any issue or any restriction on who can
+work on an issue, nor how many people can work on an issue at the same time.
+
+Contributors drive the project forward based on their own priorities and
+interests and thus you are free to work on any issue that interests you.
+
+If someone is already working on an issue that you want or need on but hasn't
```
Comment (alamb) [inline docs/source/contributor-guide/index.md:44]:
```suggestion
If someone is already working on an issue that you want or need but hasn't
```

File: docs/source/contributor-guide/index.md [hunk 2]
```diff
@@ -19,17 +19,40 @@
...
+been able to finish it yet, you should feel free to work on it as well. In
+general it is both polite and will help avoid unnecessary duplication of work if
+you leave a note on an issue when you start working on it.
+
+If you plan to work on a new feature that doesn't have an existing ticket, it is
+a good idea to open a ticket to discuss the feature. Advanced discussion often
+helps wasted effort by determining early if the feature is a good fit for
```
Comment (alamb) [inline docs/source/contributor-guide/index.md:51]:
```suggestion
helps avoid wasted effort by determining early if the feature is a good fit for
```

**Comment by simicd** [comment]
Thanks for expanding the contributor guide @alamb, very clear!

File: docs/source/contributor-guide/index.md [hunk 3]
```diff
@@ -19,17 +19,40 @@
...
+4. Code, both PR and (especially) PR Review.
 
-In addition to submitting new PRs, we have a healthy tradition of community members reviewing each other's PRs. Doing so is a great way to help the community as well as get more familiar with Rust and the relevant codebases.
+In addition to submitting new PRs, we have a healthy tradition of community
+members reviewing each other's PRs. Doing so is a great way to help the
+community as well as get more familiar with Rust and the relevant codebases.
+
+## Finding and Creating Issues to Work On
+
+You can find a curated [good-first-issue] list to help you get started.
```
Comment (jonahgao) [inline docs/source/contributor-guide/index.md:35]:
```suggestion
You can find a curated [good-first-issue](https://github.com/apache/arrow-datafusion/issues?q=is%3Aissue+is%3Aopen+label%3A%22good+first+issue%22) list to help you get started.
```


How about moving the link over here? It previously displayed as:
> 截屏2024-01-27 11 57 53

File: docs/source/contributor-guide/index.md [hunk 4]
```diff
@@ -19,17 +19,40 @@
...
+you leave a note on an issue when you start working on it.
+
+If you plan to work on a new feature that doesn't have an existing ticket, it is
+a good idea to open a ticket to discuss the feature. Advanced discussion often
+helps avoid wasted effort by determining early if the feature is a good fit for
+DataFusion before too much time is invested. It also often helps to discuss your
+ideas with the community to get feedback on implementation.
 
-You can find a curated
 [good-first-issue](https://github.com/apache/arrow-datafusion/issues?q=is%3Aissue+is%3Aopen+label%3A%22good+first+issue%22)
```
Comment (jonahgao) [inline docs/source/contributor-guide/index.md:55]:
```suggestion
```
Moved.

File: docs/source/contributor-guide/index.md (hunk 3 above)
Comment (alamb) [reply docs/source/contributor-guide/index.md:35]: Good call -- thank you @jonahgao

**Review by comphead — COMMENTED** [review]
Thanks @alamb would be nice to have a link to contribution.md from the top level page in 

https://github.com/apache/arrow-datafusion#contributing-to-datafusion section

**Comment by alamb** [comment]
> Thanks @alamb would be nice to have a link to contribution.md from the top level page in
> 
> https://github.com/apache/arrow-datafusion#contributing-to-datafusion section

Excellent call -- done in [94bbca7](https://github.com/apache/arrow-datafusion/pull/8999/commits/94bbca7f93d013d6e5336587b6b4547934544af7)

**Review by comphead — APPROVED** [review]
nice thanks @alamb

**Pull request merged**